### DISCUSSION TITLE
V5 roadmap

### DISCUSSION THREAD

**Discussion by nikhil-varma** [s-yadav/react-number-format#564] [Ideas]
Let's discuss and plan out the features/details for v5 here

**Comment by ChrisShallcross** [comment]
I've had a feature request from a user:  Is it possible to use exponential notation when adding large numbers?  For example, the user could enter 1e9 and have it "expand" to 1,000,000,000.  I've looked into this and I can't find an easy way to do this.  Is it possible?  Is it a useful feature to add?

**Reply by nikhil-varma** [reply]
Thanks for the feature request. We will look into it on it's feasibility and if possible may be consider exposing an API in the newer version

**Comment by nikhil-varma** [comment]
### Initial milestones:
- Migrate from Flow to TS
- List down all features/functions in v4

### Second milestone:
- Decide a strategy to make it headless
- If not headless, break up the functionality into importable chunks
- Implementation V5
- Create a migration strategy

### Third milestone
- Test case management/migration from v4 to v5

### Fourth milestone
- Create proper documentation site
- Clear up v4 issues

**Comment by nikhil-varma** [comment]
### Existing features
  - Formatting
    - Numeral (_Currency_ etc...)
    - Pattern (_With mask characters_)
    - Custom Formatting
    - Masking (_associated with formatting eg: '12-___-___ '_ )
    - Remove Formatting
    - Grouping styles
    - Empty Formatting
    - Separators 
       - Provide this info to Cursor Engine & Boundary Logic (_Need to be researched_) 
       - Thousands
       - Decimals
    - Leading Zeros
    - Decimal Scale customisations 
       - Fixed
       - Custom scale limit
  - Cursor Engine 
  - Cursor Boundary
     - Decouple with formatting logic 
     - Navigation through Arrow, limit cursor position depending on formatting
  - Base
     - Input element customisation for all types 
     - Custom Numerals
     - renderText
     - isAllowed
     - onValueChange
     - getInputRef
   


### Research/Unknown issues
- Pros & Cons of using Hooks vs Class 
- Handling decimals for finding boundaries (_Look for allowedDecimalSeparator_)
- Handling Negations


### Aim
- API Simplification
- Easy use-case handling by Devs (_cases apart from what library supports natively_)
- Cursor Engine (Platform dependent. Might need support for multiple platforms like RN, Vue, Angular etc.. )

**Comment by c0d3rm0n** [comment]
Do you plan to update this package to support React v19?
